core/base.py

```diff
@@ -0,0 +1,323 @@
+        relationship-type: relationship_target_id
+        """
+
+        targets_and_types = dict()
+
+        if len(relationships) > 0:
+
+            for relationship in relationships:
+                targets_and_types.update(
+                    {
+                        relationship.type:
+                            relationship.target.instance
+                            .runtime_properties.get(
+                                constants.EXTERNAL_RESOURCE_ID)
+                    }
+                )
+
+        return targets_and_types
+
+    def get_target_ids_of_relationship_type(self, relationship_type, relationships):
+
+        target_ids = []
+
+        for key, value in relationships.items():
+
+            if relationship_type in key:
+                target_ids.append(value)
+
+        return target_ids
+
+    def raise_cannot_use_external_resource(self, resource_id):
+        raise NonRecoverableError(
+            'Cannot use_external_resource because resource {0} '
+            'is not in this account.'.format(resource_id))
+
+
+class AwsBaseRelationship(AwsBase):
+
+    def __init__(self, client=None):
+        super(AwsBaseRelationship, self).__init__(client)
+        self.source_resource_id = \
+            ctx.source.instance.runtime_properties.get(
+                constants.EXTERNAL_RESOURCE_ID, None) if \
+            constants.EXTERNAL_RESOURCE_ID in \
+            ctx.source.instance.runtime_properties.keys() else \
+            ctx.source.node.properties['resource_id']
+        self.target_resource_id = ctx.target.instance.runtime_properties.get(
+                constants.EXTERNAL_RESOURCE_ID, None) if \
+            constants.EXTERNAL_RESOURCE_ID in \
+            ctx.target.instance.runtime_properties.keys() else \
+            ctx.target.node.properties['resource_id']
+        self.source_external_resource = ctx.source.node.properties['use_external_resource']
+        self.target_external_resource = ctx.target.node.properties['use_external_resource']
+        self.source_get_all_matching = None
+        self.target_get_all_matching = None
+        self.source_get_all_argument = None
+        self.target_get_all_argument = None
+
+    def associate(self):
+        return False
+
+    def associated(self):
+        if self.associate_external_resource():
+            return self.post_associate()
+        elif self.associate():
+            return self.post_associate()
+
+        raise NonRecoverableError('Unable to associate {0} with {1}.'.format(self.source_resource_id, self.target_resource_id))
+
+    def associate_external_resource(self):
+
+        if not self.source_external_resource:
+            return False
+
+        resource = self.source_get_all_matching([self.source_resource_id])[0]
+
+        if resource is None:
+            self.raise_cannot_use_external_resource(self.source_resource_id)
+
+        ctx.logger.info('not associating external resource {0}.'.format(resource.id))
```

Comment:
I'm not actually going to run  the operations, because it's external you are responsible for making sure it's associated outside of cloudify